### PR TITLE
Add EventQueue classes

### PR DESCRIPTION
**Pull request by a-veitch**
Functionality to enable doing background updates of instrumentation data (e.g. trace annotations or usage measurements) so as to minimally impact foreground task latency.

### CODE REVIEW

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    @Override
+    public void onEvent(InstrumentationEvent event, long sequence, boolean endOfBatch) {
+      event.getEntry().process();
+    }
+  }
+
+  // The single instance of the class.
+  private static final EventQueue eventQueue = new EventQueue();
+  // The event queue is built on this {@link Disruptor}.
+  private final Disruptor<InstrumentationEvent> disruptor;
```
Comment (adriancole) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: might want to somehow make disruptor optional. For example having the event queue type abstract and/or use shade to repackage disruptor. in my experience, libraries that are to be used at the lowest level of the platform easily conflict with app deps.

Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Our model is that every library (gRPC for example) will link only the API (which should not depend on any fancy library except jsr305 and guava(19).

The implementation will be linked only by the service owner. Which agree can still conflict with other libraries but it is easier to shade the disruptor (or any other dependency) when it is in only one jar.

I think having a interface makes less sense (or maybe I do not understand your point) when the implementation jar will be linked only once.

Comment (adriancole) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: sorry I wasn't meaning about an extra interface, just a base implementation (even if abstract for how to implement a part of queuing) which could allow more code reuse without a dep like this. basically batteries included but swappable (which might not be the approach you are looking for based on what you said).

Comment (adriancole) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: interestingly this type isn't used, yet (and the core-impl package is otherwise little code), so maybe just keep in mind that as the impl part grows, there will be parts others can harvest. guava is usually used by others pretty well understood etc. disruptor has a relatively low change frequency at the moment, and likely stable enough to have a core dep, but it is less understood.

Probably only async folks (who also use disruptor) who want to use this impl will need to make a custom build which shades it and publishes a variant which doesn't conflict. We can always wait until they complain (which may never happen)

File: build.gradle [hunk 2]
```diff
@@ -69,6 +69,7 @@ subprojects {
         grpcContextVersion = '1.1.2'
 
         libraries = [
+	        disruptor    : 'com.lmax:disruptor:3.3.6',
```
Comment (bogdandrutu) [inline build.gradle]: Indentation?

File: core_impl/build.gradle [hunk 3]
```diff
@@ -3,6 +3,7 @@ description = 'Instrumentation Core Impl'
 dependencies {
     compile project(':instrumentation-core'),
             project(':proto'),
+	    libraries.disruptor,
```
Comment (bogdandrutu) [inline core_impl/build.gradle]: indentation?

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    disruptor = new Disruptor<InstrumentationEvent>(new InstrumentationEventFactory(), bufferSize,
+        Executors.newSingleThreadExecutor(), ProducerType.MULTI, new SleepingWaitStrategy());
+    disruptor.handleEventsWith(new InstrumentationEventHandler());
+    disruptor.start();
+    ringBuffer = disruptor.getRingBuffer();
+  }
+
+  /**
+   * Get the {@link EventQueue} instance.
+   * @return the singleton {@code EventQueue} instance.
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Please use the formatter I suggested it will add an extra line here.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    // ensures that the event handler can take unsynchronized actions whenever possible.
+    disruptor = new Disruptor<InstrumentationEvent>(new InstrumentationEventFactory(), bufferSize,
+        Executors.newSingleThreadExecutor(), ProducerType.MULTI, new SleepingWaitStrategy());
+    disruptor.handleEventsWith(new InstrumentationEventHandler());
+    disruptor.start();
+    ringBuffer = disruptor.getRingBuffer();
+  }
+
+  /**
+   * Get the {@link EventQueue} instance.
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: s/Get/Returns

File: core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Unit tests for {@link EventQueue}.
+ */
+@RunWith(JUnit4.class)
+public class EventQueueTest {
+  /**
+   * Simple class to use that keeps an incrementing counter. Will fail with an assertion if
```
Comment (bogdandrutu) [inline core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java]: Don't need javadoc format for private class.

File: core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    private int count;
+    private volatile long id;  // stores thread ID used in first increment operation.
+
+    public Counter() {
+      count = 0;
+      id = -1;
+    }
+
+    /**
+     * Increment counter by 1. Will fail in assertion if multiple different threads are used
```
Comment (bogdandrutu) [inline core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java]: s/Increment/Increments

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+ *</pre>
+ */
+
+@ThreadSafe
+public final class EventQueue {
+  // An event in the {@link EventQueue}. Just holds a reference to an EventQueueEntry.
+  private static final class InstrumentationEvent {
+    private EventQueueEntry entry = null;
+
+    // Set the EventQueueEntry associated with this InstrumentationEvent.
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: s/Set/Sets

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+ *     int measurement = doSomeWork();
+ *     // Instead of calling update() directly, create an event to do the processing, and insert
+ *     // it into the EventQueue. It will be processed in a background thread, and doSomething()
+ *     // can return immediately.
+ *     EventQueue.getInstance.enqueue(new SomeClassUpdateEvent(this, measurement));
+ *   }
+ * }
+ *</pre>
+ */
+
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Extra empty line. Please format the code using https://github.com/google/google-java-format (if you don't use an IDE that you can configure with this formatter). For the moment use it only for the files you added.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+  // An event in the {@link EventQueue}. Just holds a reference to an EventQueueEntry.
+  private static final class InstrumentationEvent {
+    private EventQueueEntry entry = null;
+
+    // Set the EventQueueEntry associated with this InstrumentationEvent.
+    void setEntry(EventQueueEntry entry) {
+      this.entry = entry;
+    }
+
+    // Get the EventQueueEntry associated with this InstrumentationEvent.
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: s/Get/Returns

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+  // Factory for InstrumentationEvent.
+  private static final class InstrumentationEventFactory
+      implements EventFactory<InstrumentationEvent> {
+    @Override
+    public InstrumentationEvent newInstance() {
+      return new InstrumentationEvent();
+    }
+  }
+
+  /**
```
Comment (bogdandrutu) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Don't need javadoc for private class. We decided to do it only for at least package-protected classes/methods.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 1 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: More classes that use it coming soon. Re shading, lets wait and see; I suspect we may want to do this, but if so, will do a separate PR.

File: core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java (hunk 7 above)
Comment (a-veitch) [inline core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java]: Done.

File: core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java (hunk 6 above)
Comment (a-veitch) [inline core_impl/src/test/java/com/google/instrumentation/common/EventQueueTest.java]: Done

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 4 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 5 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 11 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 10 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 8 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java (hunk 9 above)
Comment (a-veitch) [inline core_impl/src/main/java/com/google/instrumentation/common/EventQueue.java]: Done.

File: core_impl/build.gradle (hunk 3 above)
Comment (a-veitch) [inline core_impl/build.gradle]: Another TAB :-(

File: build.gradle (hunk 2 above)
Comment (a-veitch) [inline build.gradle]: Looks like a TAB snuck in there.

**Comment by a-veitch** [comment]
PTAL

**Comment by adriancole** [comment]
SGTM
